### PR TITLE
DOC: Adds transfer learning tutorial

### PR DESCRIPTION
**Pull request by thomasjpfan**
This addresses https://github.com/dnouri/skorch/issues/261 and maybe https://github.com/dnouri/skorch/issues/278.

I covered finetuning and how to use a custom train/valid dataset. I think its pretty common for people coming from the pure `PyTorch` approach to want to provide their own train/validation datasets.

I approached the custom datasets by defining a custom `train_valid` function:

```python
def train_valid_split(train, valid, **kwargs):
    return train, valid

net = NeuralNetClassifier(..., train_split=train_valid_split)
net.fit(train_ds, val_ds);
```

This is a pretty hacky approach to the problem. By defining `train_valid_split ` that just returns its inputs, I treated `X` as the training dataset, and `y` as the validation set.

### CODE REVIEW

**Comment by thomasjpfan** [comment]
The accuracy is on par with the original tutorial. It’s the benefit of using a pretrained model and only training the final layers.

I would want this tutorial to only have best practices. I think not having an official way to support custom training and validation datasets, is a high barrier of entry.

**Comment by benjamin-work** [comment]
> It’s the benefit of using a pretrained model and only training the final layers

It works really well.

> I think not having an official way to support custom training and validation datasets, is a high barrier of entry.

Yes, that would be ideal. When we started out with skorch, we put a lot of thought into how to manage train/test splits and even though the current solution does not support this use case, we found it to be the best. At the same time, we thought about implementing a "predefined split" that would support this exact use case, but we left it open. Maybe now is the time to implement it.

I guess that it could look something like this:

```python
class PredefinedSplit:
    def __init__(self, dataset_valid):
        self.dataset_valid = dataset_valid

    def __call__(self, dataset, y=None, groups=None):
        if self.dataset_valid is None:
            raise ValueErrror("some useful message")
        return dataset, self.dataset_valid

    def __getstate__(self):
        state = self.__dict__.copy()
        state['dataset_valid'] = None
        return state
```
But there are certainly some edge cases that still need covering.

**Comment by taketwo** [comment]
> Maybe now is the time to implement it.

:+1: from my side. I work mostly with datasets that have official train/test splits. So far I've resorted to using custom scoring functions that capture test dataset and ignore the data that `EpochScoring` passes to them. But this is again hacky and it would be great to have skorch support for predefined splits.

**Comment by benjamin-work** [comment]
I created a new issue: #282

Regarding this PR: It could either be left as is until the new feature is included, or we wait for it first.

**Comment by thomasjpfan** [comment]
1. I will wait for #282 to be resolved.

2. Regarding `examples` vs `notebooks`, as long as I can keep all the logic in one notebook, I would prefer this tutorial to be in `notebooks`. 

3. I added a `datasets` directory into `notebooks` that can be used by notebooks. The downloading logic is placed in the notebook itself. This allows the notebook to be able to run on all platforms.

**Comment by benjamin-work** [comment]
This looks really good. I have only 2 minor points left:

The download function could be made a little more flexible by allowing to pass the folder. I made this and some minor adjustments with when/how to print things:

```
def download_and_extract_data(dataset_dir='datasets'):
    data_zip = os.path.join(dataset_dir, 'hymenoptera_data.zip')
    data_path = os.path.join(dataset_dir, 'hymenoptera_data')
    url = "https://download.pytorch.org/tutorial/hymenoptera_data.zip"

    if not os.path.exists(data_path):
        if not os.path.exists(data_zip):
            print("Starting to download data...")
            data = request.urlopen(url, timeout=15).read()
            with open(data_path, 'wb') as f:
                f.write(data)

        print("Starting to extract data...")
        with ZipFile(data_zip, 'r') as zip_f:
            zip_f.extractall(dataset_dir)
        
    print("Data has been downloaded and extracted to {}.".format(dataset_dir))
```

Also, on my machine, the final validation accuracy was a little lower than 0.9608, so maybe you could adjust the sentence in the last paragraph to something like `...with a validiation accuracy of roughly 0.96`.

**Comment by BenjaminBossan** [comment]
Looks good, fantastic work. I will merge on Monday.

File: notebooks/Transfer_Learning.ipynb [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+   "metadata": {},
+   "source": [
+    "# Transfer Learning with skorch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, you will learn how to train a neutral network using transfer learning with the `skorch` API. Transfer learning uses a pretrained model to initialize a network. This tutorial converts the pure Pytorch approach in [Pytorch's Transfer Learning tutorial](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html#sphx-glr-beginner-transfer-learning-tutorial-py) into using `skorch`.\n",
```
Comment (benjamin-work) [inline notebooks/Transfer_Learning.ipynb]: > This tutorial converts the pure Pytorch approach in ... **into** using `skorch`

There is a word missing after `into`.

**Comment by benjamin-work** [comment]
I was just about to merge when I discovered a small error (see comment). After this, we should really be good.

File: notebooks/Transfer_Learning.ipynb (hunk 1 above)
Comment (taketwo) [reply notebooks/Transfer_Learning.ipynb]: Nitpick: we typically capitalize PyTorch.

Comment (thomasjpfan) [reply notebooks/Transfer_Learning.ipynb]: I reworded the sentence into the following:

This tutorial converts the pure PyTorch approach described in [PyTorch's Transfer Learning Tutorial](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html) to `skorch`.
